src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+        // *member-access* or *pointer-member-access*, even if there is no other possible parse of the
+        // sequence of tokens. Note that these rules are not applied when parsing a *type-argument-list*
+        // in a *namespace-or-type-name* (§3.8).
+        //
+        private ExpressionSyntax ParseExpressionOrDeclaration(ParseTypeMode mode, MessageID feature, bool permitTupleDesignation)
+        {
+            if (IsPossibleDeclarationExpression(mode, permitTupleDesignation))
+            {
+                TypeSyntax type = this.ParseType(mode);
+                var designation = ParseDesignation();
+                if (feature != 0)
+                {
+                    designation = CheckFeatureAvailability(designation, feature);
+                }
+
+                return _syntaxFactory.DeclarationExpression(type, designation);
+            }
+            else
+            {
+                return this.ParseSubExpression(Precedence.Expression);
+            }
+        }
+
+        private bool IsPossibleDeclarationExpression(ParseTypeMode mode, bool permitTupleDesignation)
+        {
+            if (this.IsInAsync && this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+            {
+                // can't be a declaration expression.
+                return false;
+            }
+
+            var resetPoint = this.GetResetPoint();
+            try
+            {
+                SyntaxToken lastTokenOfType;
+                switch (ScanType(out lastTokenOfType))
+                {
+                    case ScanTypeFlags.PointerOrMultiplication:
+                        if (mode == ParseTypeMode.FirstElementOfPossibleTupleLiteral || mode == ParseTypeMode.AfterTupleComma)
+                        {
+                            // Tuples cannot contain pointer types because pointers may not be generic type arguments.
+                            return false;
+                        }
+                        break;
+
+                    case ScanTypeFlags.NotType:
+                        return false;
+                }
+
+                // check for a designation
+                if (!ScanDesignation(permitTupleDesignation && (lastTokenOfType.IsVar() || IsPredefinedType(lastTokenOfType.Kind))))
+                {
+                    return false;
+                }
+
+                return mode != ParseTypeMode.FirstElementOfPossibleTupleLiteral || this.CurrentToken.Kind == SyntaxKind.CommaToken;
+            }
+            finally
+            {
+                this.Reset(ref resetPoint);
+                this.Release(ref resetPoint);
+            }
+        }
+
+        private bool ScanDesignation(bool permitTuple)
+        {
+            switch (this.CurrentToken.Kind)
+            {
+                default:
+                    return false;
+                case SyntaxKind.IdentifierToken:
+                    var result = this.IsTrueIdentifier();
+                    this.EatToken();
```

Comment:
I think it would be cleaner and more intuitive if this function (and the like) wouldn't consume tokens  it doesn't accept.